scripts/globals/combat/physical_utilities.lua

```diff
@@ -522,6 +522,108 @@ xi.combat.physical.calculateRangedPDIF = function(actor, target, weaponType, wsA
+        statBonus = 0.15
+    elseif dDex >= 40 then
+        statBonus = (dDex - 35) / 100
+    elseif dDex >= 30 then
+        statBonus = 0.04
+    elseif dDex >= 20 then
+        statBonus = 0.03
+    elseif dDex >= 14 then
+        statBonus = 0.02
+    elseif dDex >= 7 then
+        statBonus = 0.01
+    end
+
+    return statBonus
+end
+
+-- Innin: Critical hit rate bonus when actor is behind target.
+xi.combat.physical.criticalRateFromInnin = function(actor, target)
+    local inninBonus = 0
+
+    if
+        actor:hasStatusEffect(xi.effect.INNIN) and
+        actor:isBehind(target, 23)
+    then
+        inninBonus = actor:getStatusEffect(xi.effect.INNIN):getPower()
+    end
+
+    return inninBonus
+end
+
+-- Fencer: Critical hit rate bonus when actor is only wielding with main hand.
+xi.combat.physical.criticalRateFromFencer = function(actor)
+    local fencerBonus = 0
+
+    local mainEquip = actor:getStorageItem(0, 0, xi.slot.MAIN)
+    local subEquip  = actor:getStorageItem(0, 0, xi.slot.SUB)
+
+    if
+        actor:getObjType() == xi.objType.PC and
+        mainEquip and
+        not mainEquip:isTwoHanded() and                                                      -- No 2 handed weapons.
+        not mainEquip:isHandToHand() and                                                     -- No 2 handed weapons.
+        (subEquip == nil or subEquip:getSkillType() == xi.skill.NONE or subEquip:isShield()) -- Only shields allowed in sub.
+    then
+        fencerBonus = actor:getMod(xi.mod.FENCER_CRITHITRATE) / 100
+    end
+
+    return fencerBonus
+end
+
+-- Critical rate from Building Flourish.
+-- TODO: Study case were if we can attach modifiers to the effect itself, both this and the effect may need refactoring.
+xi.combat.physical.criticalRateFromFlourish = function(actor)
+    local buildingFlourishBonus = 0
+
+    if actor:hasStatusEffect(xi.effect.BUILDING_FLOURISH) then
+        local effectPower    = actor:getStatusEffect(xi.effect.BUILDING_FLOURISH):getPower()
+        local effectSubPower = actor:getStatusEffect(xi.effect.BUILDING_FLOURISH):getSubPower()
+
+        if effectPower >= 3 then
+            buildingFlourishBonus = (10 + effectSubPower) / 100
+        end
+    end
+
+    return buildingFlourishBonus
+end
+
+-- Critical rate master function.
+xi.combat.physical.calculateSwingCriticalRate = function(actor, target, isWeaponskill, TP1000, TP2000, TP3000)
+    -- See reference at https://www.bg-wiki.com/ffxi/Critical_Hit_Rate
+    local finalCriticalRate     = 0
+    local baseCriticalRate      = 0.05
+    local statBonus             = xi.combat.physical.criticalRateFromStatDiff(actor, target)
+    local inninBonus            = xi.combat.physical.criticalRateFromInnin(actor, target)
+    local fencerBonus           = xi.combat.physical.criticalRateFromFencer(actor)
+    local buildingFlourishBonus = xi.combat.physical.criticalRateFromFlourish(actor)
+    local modifierBonus         = actor:getMod(xi.mod.CRITHITRATE) / 100
```

Comment:
cpp has `ENEMYCRITRATE` mod as well which has a comment of `ENEMYCRITRATE            = 166, // Raises chance enemy will crit` which would appear to be additive to this modifierBonus